packages/puppeteer-core/src/common/util.ts

```diff
@@ -522,26 +523,12 @@ export async function waitWithTimeout<T>(
   taskName: string,
   timeout: number
 ): Promise<T> {
-  let reject: (reason?: Error) => void;
-  const timeoutError = new TimeoutError(
-    `waiting for ${taskName} failed: timeout ${timeout}ms exceeded`
-  );
-  const timeoutPromise = new Promise<never>((_, rej) => {
-    return (reject = rej);
+  const deferred = createDeferred<never>({
+    message: `waiting for ${taskName} failed: timeout ${timeout}ms exceeded`,
+    timeout,
   });
-  let timeoutTimer = null;
-  if (timeout) {
-    timeoutTimer = setTimeout(() => {
-      return reject(timeoutError);
-    }, timeout);
-  }
-  try {
-    return await Promise.race([promise, timeoutPromise]);
-  } finally {
-    if (timeoutTimer) {
-      clearTimeout(timeoutTimer);
```

Comment:
yeah we will likely need to to reject the deferred here if the race succeed.

Reply:
Waiting for other PRs to rebase and push the update.